### PR TITLE
Add field groups functionality

### PR DESCRIPTION
**Pull request by delano**
### **User description**
Adds field grouping functionality to organize related fields in Familia models. Features can now declare their own groups, and users can define custom groups to categorize fields by purpose.

The implementation maintains backward compatibility with existing feature APIs (transient_fields, encrypted_fields) while providing a new unified field_groups interface. Fields defined within a field_group block are automatically tracked in that group, and features that declare a field_group parameter get their own group initialized.

Key changes include adding field_group support to the FeatureDefinition data class, automatic group initialization for features, and preventing nested field groups. The 197-line test suite covers manual groups, feature groups, mixed scenarios, inheritance, and error cases.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude <noreply@anthropic.com>

___

### **PR Type**
Enhancement

___

### **Description**
- Add field groups functionality to organize related fields

- Integrate field groups with existing features (transient, encrypted)

- Maintain backward compatibility with existing APIs

- Add comprehensive test coverage with 197-line test suite

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["FeatureDefinition"] -- "add field_group param" --> B["Feature System"]
  B -- "initialize groups" --> C["Field Groups"]
  C -- "track fields" --> D["Transient Fields"]
  C -- "track fields" --> E["Encrypted Fields"]
  C -- "manual grouping" --> F["Custom Groups"]
  F -- "prevent nesting" --> G["Validation"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>base.rb</strong><dd><code>Add field_group parameter to feature system</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/familia/base.rb

<ul><li>Add <code>field_group</code> parameter to <code>add_feature</code> method<br> <li> Update <code>FeatureDefinition</code> instantiation with field_group support<br> <li> Maintain backward compatibility with existing feature system</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/142/files#diff-adc0a2d820ca9b2723e3508e09df80ff20f72cc0a1daebeeb33936429de556a3">+6/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>features.rb</strong><dd><code>Extend FeatureDefinition with field_group support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/familia/features.rb

<ul><li>Update <code>FeatureDefinition</code> data class to include <code>field_group</code> attribute<br> <li> Add automatic field group initialization for features that declare <br>groups<br> <li> Integrate field group creation with feature loading process</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/142/files#diff-d49db178a8019ef15e9a9607f309f745b17d882ad39f4535167dd90320eb2f13">+6/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>encrypted_fields.rb</strong><dd><code>Integrate encrypted fields with field groups</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/familia/features/encrypted_fields.rb

<ul><li>Add <code>field_group: :encrypted_fields</code> to feature registration<br> <li> Integrate encrypted fields with field groups system<br> <li> Automatically add encrypted fields to their group when defined</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/142/files#diff-efdfda22d12bc384dbf33527e9fef37ba7e3d353ffdcfa4db18fda6f71ecc1fd">+7/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>transient_fields.rb</strong><dd><code>Integrate transient fields with field groups</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/familia/features/transient_fields.rb

<ul><li>Add <code>field_group: :transient_fields</code> to feature registration<br> <li> Integrate transient fields with field groups system<br> <li> Automatically add transient fields to their group when defined</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/142/files#diff-0ae4ca9f7b1deead120ca916fe46e92169d6bc4fda74c7b0d31620aa9217f3f8">+8/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>definition.rb</strong><dd><code>Implement core field groups functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/familia/horreum/definition.rb

<ul><li>Add <code>field_group</code> method for manual field grouping with block support<br> <li> Add <code>field_groups</code> accessor method returning hash of groups<br> <li> Implement nested field group prevention with validation<br> <li> Integrate field group tracking with field registration system</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/142/files#diff-5d439fbd77f17e56a3c1ac01218164d4239a7a9f9b291cab396fceeb8e008174">+85/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>field_groups_try.rb</strong><dd><code>Add comprehensive field groups test suite</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

try/features/field_groups_try.rb

<ul><li>Add comprehensive test coverage for field groups functionality<br> <li> Test manual groups, feature groups, mixed scenarios, and inheritance<br> <li> Include error cases and edge cases validation<br> <li> Verify backward compatibility with existing APIs</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/142/files#diff-612e71c76008b8bab0aea79eda2c8eb561ad2f15331d191c2cbf49cef3c8237d">+209/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: lib/familia/features/transient_fields.rb [hunk 1]
```diff
@@ -104,12 +104,14 @@ module Features
...
     module TransientFields
-      Familia::Base.add_feature self, :transient_fields, depends_on: nil
+      Familia::Base.add_feature self, :transient_fields, depends_on: nil, field_group: :transient_fields
 
       def self.included(base)
         Familia.trace :LOADED, self, base if Familia.debug?
         base.extend ModelClassMethods
 
+        require_relative 'transient_fields/transient_field_type'
+
```
[the remark this answers is not in the document]
Comment (delano) [reply lib/familia/features/transient_fields.rb:114]: It was moved from 154, not duplicated.
